在paddle.tensor中加入float，int和complex等数据转换接口

Qin-sx opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->

User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->

Improvements

### Description
<!-- Describe what you’ve done -->

**当前问题**  
paddle中没有float16和bfloat16等数据转换接口，需要用astype转换

**解决方案**
注册float16和bfloat16等数据转换接口

Qin-sx committed:
added float16 and bfloat16 in tensor

	modified:   python/paddle/base/dygraph/math_op_patch.py
	new file:   test/legacy_test/test_tensor_float16_bfloat16.py

zhwesky2010 commented on python/paddle/base/dygraph/math_op_patch.py:
```diff
@@ -104,6 +104,26 @@ def astype(self: Tensor, dtype: DTypeLike) -> Tensor:
 
         return _C_ops.cast(self, dtype)
 
+    def float16(self: Tensor) -> Tensor:
```
这个是不是得整体加，包括fp32、fp64、int32等所有类型，而不是给某两个类型增加一个函数。

实现方式建议通过一个公共函数，批量注册

Qin-sx replied:
嗯，收到，已修改为通过公共函数批量注册

Qin-sx committed:
Merge branch 'develop' into tensor_float16_bfloat16

Qin-sx committed:
added convert for all types

	modified:   python/paddle/base/dygraph/math_op_patch.py
	modified:   test/legacy_test/test_tensor_float16_bfloat16.py

Qin-sx committed:
rename the test

	renamed:    test/legacy_test/test_tensor_float16_bfloat16.py -> test/legacy_test/test_tensor_type_convert_api.py

Qin-sx commented:
数据类型参考 https://www.paddlepaddle.org.cn/documentation/docs/api/paddle/Tensor_cn.html#dtype

Qin-sx committed:
added pir modification

	modified:   python/paddle/pir/math_op_patch.py
	modified:   test/legacy_test/test_tensor_type_convert_api.py

zhwesky2010 commented on python/paddle/base/dygraph/math_op_patch.py:
```diff
@@ -64,6 +64,23 @@
+    'int32': 'int32',
+    'int64': 'int64',
+    # other
+    'bool': 'bool',
```
这个应该还有一些其他的用法吧，或者别名用法，也需要配置一下：
x.long、x.double、x.cfloat、x.uint8

这个要配全

Qin-sx replied:
嗯，收到，已修改，之前是按照上面的文档的tensor type写的

zhwesky2010 commented on python/paddle/pir/math_op_patch.py:
```diff
@@ -37,6 +37,23 @@
     DataType.INT64,
 ]
 
+_supported_dtype_conversions = {
```
同上

Qin-sx replied:
已修改

zhwesky2010 commented on test/legacy_test/test_tensor_type_convert_api.py:
```diff
@@ -0,0 +1,217 @@
+    Unit tests for all supported tensor dtype conversion methods.
+    """
+
+    _supported_dtype_conversions = {
```
同上

Qin-sx replied:
已修改

Qin-sx committed:
added more types

	modified:   python/paddle/base/dygraph/math_op_patch.py
	modified:   python/paddle/pir/math_op_patch.py
	modified:   test/legacy_test/test_tensor_type_convert_api.py

zhwesky2010 reviewed (approved):
LGTM

Qin-sx committed:
Merge branch 'develop' into tensor_float16_bfloat16

luotao1 commented:
@Qin-sx xpu-test 一直失败
